### ISSUE TITLE
running node coap server leaks on receipt of coap request

### ISSUE THREAD

**Issue by username-II** [mcollina/node-coap#135]
'use strict';
var coap =require('coap');

// actual server code
var localServer =coap.createServer(undefined,function(i,o){
  o.on('error', function(err){
    console.log('coap server error: '+err.toString());
  });
  o.on('timeout', function(){
    console.log('coap server reply timeout.');
  });
  console.log('coap server uri: '+i.url);
  console.log('memory: '+JSON.stringify(process.memoryUsage()));
  return o.end();
});

// create and decode client requests
function sendQuery(func,machine,port,path,args,body,done){
  var req =coap.request({ 
    method:func, 
    hostname:machine, 
    port:port, observe:false, 
    pathname:path, query:args
  });
  req.on('error', function(err){
    console.log('coap query error: '+func+' '+machine+':'+port+path+'&'+args+' >'+err.toString());
    done(undefined);
  });
  req.on('timeout', function(){
    // error callback above is called first
    console.log('coap query timeout: '+func+' '+machine+':'+port+path+'&'+args);
  });
  req.on('response', function(res){
    console.log('coap responce to ['+func+' '+machine+':'+port+path+'&'+args+']: ' +res.code);
    var obj =null; //cborToObj(res.payload);
    done(obj);
  }.bind(this));
  // done
  req.end();
  console.log('coap query sent: '+func+' '+machine+':'+port+path+'&'+args);
};

// the coap callback
var cb =function(o){console.log('body:' +JSON.stringify(o))};

// uses arg2 as <server port>
var port =parseInt(process.argv[3], 10);
if(port <0 || port >65535 || typeof port != 'number' || isNaN(port)) port =5683;

// if arg1 is 'server' start a server
if( process.argv[2] =='server' ){
  console.log('create server port '+port);
  localServer.listen(port);
} else {
  // test spam the server
  var postsPerSec =100;
  setInterval(function(){
    // tried with PUT, POST, GET etc...
    sendQuery('POST','127.0.0.1',port,'/well-known/core','',undefined,cb);
  }, 1000/ postsPerSec);
}

// run: node minServeLeak.js server
// watch the above process memory use
// run: node minServeLeak.js
// watch server heap total and heap used memory rise and never fall...

// tried with node v4.4.7 lts & v6.3.9 lts 32b & 64b on win10 x64 & win8.1 x64

// attempted node args to force gc but leak still appears:
// node.exe --min_progress_during_incremental_marking_finalization=1 --max_incremental_marking_finalization_rounds=4 --optimize_for_size --escape_analysis_iterations=8 --always_compact --gc_global --gc_interval=1 --expose_gc_as=v8gc --max_executable_size=48 minServeLeak.js server

**Comment by mcollina** [comment]
how many requests are you running?

**Comment by username-II** [comment]
one console running it in server mode and showing leak: node minServeLeak.js server
one other console running it in client spam mode: node minServeLeak.js
client mode continuously spams at ~100 posts per second. From src:
...
var postsPerSec =100;
setInterval(function(){
// tried with PUT, POST, GET etc...
sendQuery('POST','127.0.0.1',port,'/well-known/core','',undefined,cb);
}, 1000/ postsPerSec);
...

**Comment by username-II** [comment]
Note: Running it slower just slows down the leak.

**Comment by mcollina** [comment]
How big does it gets?
There is an LRU cache defined here https://github.com/mcollina/node-coap/blob/667d1a6cd3588bb2c20d58edaf4ff346da60b45a/lib/server.js#L80-L94

**Comment by username-II** [comment]
I got bored running the test ~@0.75GB but I have seen it get larger and eventually node.exe quits out with out of mem & failed alloc.
Certainly far more then the 33554432 bytes you have allocated for the cache.
The messages are all processed fine, until the leak causes memory use majorly hitting the swap and/or GC.
This then eventually causes a processing delay longer then the cache you point at can support.
I don't think the cache needs enlarging; but the leak seems to be somewhere else.

**Comment by username-II** [comment]
If the leak was the cache then I would expect it to plateau eventually. I am not seeing this.

**Comment by mcollina** [comment]
I do not have time to debug & fix this issues as per #97. I will be happy to review a pull request that fixes it.
the new `--inspect` flag and doing heap diffs will tell you what is leaking.

**Comment by username-II** [comment]
I thank you for your time.
I did go down down the route of using '--inspect' to try and find the issue but my node & javascript leak-fu is rather rudimentary.
I shall give it another go.

**Comment by username-II** [comment]
After further lookage by @richardms it definitly seems to be caused by the lru-cache not shrinking. After a little testing and mangling of @richardms solution by myself I have the following mitigations to the above posted code to remove the leak:

// ensure the lru-cache buffer is pruned regulary, cpu use here seems fine for my needs
var localServer =coap.createServer(undefined,function(i,o){
  localServer._lru.prune(); // <- fix leak on coap dgram recieve
...
};

Also as a possibile optimisation:

// Additionally override the lru-cache message length for a better use of cache
function makeSymbol(k) { return (typeof Symbol === 'function') ?Symbol.for(k) :'_'+k; }
localServer._lru[makeSymbol('lengthCalculator')] =function(n){ return n.buffer.byteLength; };

**Comment by neophob** [comment]
@username-II your workaround is not really the best way to solve this issue. i think what make sense is add a ttl (like 1 minute) to the lru cache items, so they get cleaned up automatically.

**Comment by richardms** [comment]
Hi all,

I think there are actually two issues here, the main one being that the length function passed to the LRU constructor vastly underestimates the length of the cached data. This is because it's returning the amount of the packet buffer that's used, which is in the order of bytes, not the underlying buffer size (8192).  Altering the length function to return the buffer length improves matters (but ignores any object storage overhead):

```
  this._lru = LRU({
      max: options.cacheSize || (32768 * 1024)
    , length: function(n) {
        var buffer = n.buffer;
        return buffer.byteLength
      }
    , maxAge: parameters.exchangeLifetime
    , dispose:  function(key, value) {
                  if (value.sender)
                    value.sender.reset()
                }
  })
```

The second issue is the lack of pruning when the memory limit isn't hit - which means the server can suffer from bloat, retaining very old packets. So I like @neophob's idea to `get` the value explicitly from the lru after a ttl - which would act to remove it.

Do you want me to create a pull-request?

**Comment by mcollina** [comment]
@richardms please go ahead.

**Comment by richardms** [comment]
OK. I've created the pull-request, please take a look.

I've also changed the maxAge figure as the LRU cache expects ms not s.

**Comment by mcollina** [comment]
Fixed in v0.19.0.
